runtime/lua/vim/filetype.lua

```diff
@@ -2297,7 +2603,13 @@ function M.match(arg)
   if contents then
     -- Sanity check: this should not happen
     assert(not bufnr, '"buf" and "contents" are mutually exclusive')
-    -- TODO: "scripts.lua" content matching
+    -- If name is nil, catch any errors from the contents filetype detection function.
+    -- If the function tries to use the filename that is nil then it will fail,
+    -- but this enables checks which do not need a filename to still work.
+    ft = pcall(match_contents, contents, name)
+    if ft then
+      return ft
+    end
```

Comment:
On further consideration (and some poking from @clason) this block should be moved to the end of `match()` and should not require that `bufnr` is nil. Basically, even when given a valid `bufnr`, if `match()` fails to find a filetype using the "standard" methods (filename, pattern, extension), it should then fall back to doing full content-based detection.

Reply:
(but using `contents` instead of the buffer contents if that was passed)

Reply:
and (sorry 😭 ) then the `match_from` functions can be moved (back...) to `detect.lua`?

Reply:
> and (sorry sob ) then the `match_from` functions and patterns can be moved (back...) to `detect.lua`?

Does that include all patterns (`patterns_no_hashbang` and `patterns_hashbang`) or only the more complex checks that use a function? After all, the extension and filename patterns are in `filetype.lua` also.

Reply:
I'd say the patterns as well (which are _content_ patterns as opposed to _filename_  patterns) -- the goal is "locality": keeping related things close together, so if the patterns are only used by a function in `detect.lua`, they should be in `detect.lua`?

Does that make sense?